src/silx/io/dictdumplinks.py

```diff
@@ -0,0 +1,462 @@
+"""
+Supports representations of these HDF5 link types:
+
+* Soft links: `h5py.SoftLink`
+* External links: `h5py.ExternalLink`
+* Virtual datasets: `h5py.VirtualLayout`
+* External raw binary data: `ExternalBinaryLayout`
+
+An instance of these link classes can be created from a serialized
+link with `link_from_serialized` or from an HDF5 item with `link_from_hdf5`.
+
+Serialized instances of HDF5 links have the following type
+
+* Soft links: string or `DataUrl`
+* External links: string or `DataUrl`
+* Virtual datasets: dictionary matching `VirtualDatasetSchemaV1`
+* External raw binary data: dictionary matching `ExternalBinarySchemaV1`
+"""
+
+import os
+from typing import NewType, TypedDict, Any, Literal
+from collections.abc import Sequence, Mapping
+
+import h5py
+import numpy
+import fabio
+from fabio.fabioimage import FabioFrame
+from fabio.edfimage import EdfFrame
+from fabio.TiffIO import TiffIO
+from numpy.typing import DTypeLike
+
+from .url import DataUrl
+
+RawDsetIndexItem = NewType("RawDsetIndexItem", int | slice | type(Ellipsis) | None)
+RawDsetIndex = NewType("RawDsetIndex", tuple[RawDsetIndexItem, ...] | RawDsetIndexItem)
+
+SliceArgs = NewType(
+    "SliceArgs",
+    Sequence[int | None]  # slice(None, stop, None)
+    | Sequence[int | None, int | None]  # slice(start, stop)
+    | Sequence[int | None, int | None, int | None],  # slice(start, stop, step)
+)
+DsetIndexItem = NewType("DsetIndexItem", RawDsetIndexItem | SliceArgs)
+DsetIndex = NewType("DsetIndex", Sequence[DsetIndexItem] | DsetIndexItem)
+
+
+class VdsSourceSchema(TypedDict):
+    file_path: str
+    data_path: str
+    shape: Sequence[int]
+    source_index: DsetIndex
+    target_index: DsetIndex
+
+
+class VirtualDatasetSchemaV1(TypedDict):
+    dictdump_schema: Literal["virtual_dataset_v1"]
+    shape: Sequence[int]
+    dtype: str | DTypeLike
+    sources: (
+        Sequence[VdsSourceSchema]
+        | VdsSourceSchema
+        | Sequence[h5py.VirtualSource]
+        | h5py.VirtualSource
+    )
```

Comment:
@t20100 @payno  I'm using `TypedDict` but this is just for documentation, not for validation or type coercion. For that we would need `pydantic` but I guess we won't add that as a base dependency to silx?